content/codes/components/enigma-path-svg.ts

```diff
@@ -0,0 +1,558 @@
+/*
+ *  SVG interactive Enigma view including encoding path animation
+ */
+
+import {Machine} from './enigma';
+import {$, $$, ElementView} from '@mathigon/boost';
+
+interface EnigmaPathSVG {
+  drawMachine: (encodingpath: any) => void;
+  drawPath: (encodingpath: string[], animate: boolean) => void;
+}
+
+interface SVGAnimateMotionElement extends SVGAnimationElement {
+  beginElement(): SVGAnimateMotionElement;
+}
+
+/*
+  On a keypress drawPath will be called and the encoded path returned via the callback
+*/
+export function createEnigmaPathSVG(svgid: string, machine: Machine, keypresscallback:(k:string[]) => void) : any {
+  const xmlns = 'http://www.w3.org/2000/svg';
+
+  // TODO remove duplication across path and machine source
+  const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
+
+  function charToIndex(ch:string) {
+    const charA = 'A'.charCodeAt(0);
+    return ch.charCodeAt(0) - charA;
+  }
+
+  // Some simple utility methods
+  function getYPos(n:string, off:number) {
+    const pos = (charToIndex(n) + off + 1) % letters.length;
+    return pos ? pos : letters.length;
+  }
+
+  function addText(layer: ElementView, text: string, x: number, y: number, cls: string) {
+    layer.html += '<text text-anchor="middle" alignment-baseline="middle" x="' +
+      x + '" y="' + y + '" class="' + cls + '">' + text + '</text>';
+  }
+
+  function makept(x:number, y:number) {
+    return {x, y};
+  }
+
+  function move(x:number, y:number) {
+    return 'M' + x + ',' + y + ' ';
+  }
+  function line(x:number, y:number) {
+    return 'L' + x + ',' + y + ' ';
+  }
+  function arc(r:number, x:number, y:number) {
+    return 'A' + r + ',' + r + ' 0 0 1 ' + x + ',' + y + ' ';
+  }
+
+  function path(p: string, width: number, stroke: string, fill: string, id?: string, dasharray?: number[], dashoffset?: number) {
+    return '<path ' + (id ? ' id="' + id + '"' : '') +
+      ' d="' + p + '" stroke-width=' + width + ' stroke="' + stroke +
+      (dasharray ? '" stroke-dasharray="' + dasharray.join(' ') : '') +
+      (dashoffset ? '" stroke-dashoffset="' + dashoffset : '') +
+      '" fill="' + fill + '"></path>';
+  }
+
+  function pathpts(points: any[], close = false) {
+    return 'M' + points.map(p => p.x + ',' + p.y).join('L') + (close ? 'Z' : '');
+  }
+
+  function addRoundRect(layer:ElementView, x:number, y:number, width:number, height:number, rl:number, rr:number,
+      stroke:string, fill:string) {
+    let path = '<path d="';
+    path += move(x + rl, y);
+    path += line(x + width - rr, y);
+    path += arc(rr, x + width, y + rr);
+    path += line(x + width, y + height - rr);
+    path += arc(rr, x + width - rr, y + height);
+    path += line(x + rl, y + height);
+    path += arc(rl, x, y + height - rl);
+    path += line(x, y + rl);
+    path += arc(rl, x + rl, y);
+    path += '" stroke-width="1" stroke="' + stroke + '" fill="url(#' + fill + ')" />';
+    layer.html += path;
```

Comment:
You can specify rounded corners using the SVG <rect> element, e.g. something like 
```ts
$N('rect', {x, y, width, height, r}, layer);
```

Reply:
But you can't have one side rounded and one side square - which your mockup had for the reflector, so I thought it might be needed.  If not, then this is much easier.